src/a2dp-lhdc.c

```diff
@@ -0,0 +1,460 @@
+ *
+ * This project is licensed under the terms of the MIT license.
+ *
+ */
+
+#include "a2dp-lhdc.h"
+
+#include <errno.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <glib.h>
+
+#include <lhdcBT.h>
+#include <lhdcBT_dec.h>
+
+#include "a2dp.h"
+#include "audio.h"
+#include "ba-transport-pcm.h"
+#include "bluealsa-config.h"
+#include "io.h"
+#include "rtp.h"
+#include "utils.h"
+#include "shared/a2dp-codecs.h"
+#include "shared/defs.h"
+#include "shared/ffb.h"
+#include "shared/log.h"
+#include "shared/rt.h"
+
+static const struct a2dp_channel_mode a2dp_lhdc_channels[] = {
+	{ A2DP_CHM_STEREO, 2, LHDC_CHANNEL_MODE_STEREO },
+};
+
+static const struct a2dp_sampling_freq a2dp_lhdc_samplings[] = {
+	{ 44100, LHDC_SAMPLING_FREQ_44100 },
+	{ 48000, LHDC_SAMPLING_FREQ_48000 },
+	{ 96000, LHDC_SAMPLING_FREQ_96000 },
+};
+
+struct a2dp_codec a2dp_lhdc_sink = {
+	.dir = A2DP_SINK,
+	.codec_id = A2DP_CODEC_VENDOR_LHDC_V3,
+	.capabilities.lhdc_v3 = {
+		.info = A2DP_SET_VENDOR_ID_CODEC_ID(LHDC_V3_VENDOR_ID, LHDC_V3_CODEC_ID),
+		.frequency =
+			LHDC_SAMPLING_FREQ_44100 |
+			LHDC_SAMPLING_FREQ_48000 |
+			LHDC_SAMPLING_FREQ_96000,
+		.bit_depth =
+			LHDC_BIT_DEPTH_16 |
+			LHDC_BIT_DEPTH_24,
+		.jas = 0,
+		.ar = 0,
+		.version = LHDC_VER3,
+		.max_bit_rate = LHDC_MAX_BIT_RATE_900K,
+		.low_latency = 0,
+		.llac = 1,
+		.ch_split_mode = LHDC_CH_SPLIT_MODE_NONE,
+		.meta = 0,
+		.min_bitrate = 0,
+		.larc = 0,
+		.lhdc_v4 = 1,
+	},
+	.capabilities_size = sizeof(a2dp_lhdc_v3_t),
+	.channels[0] = a2dp_lhdc_channels,
+	.channels_size[0] = ARRAYSIZE(a2dp_lhdc_channels),
+	.samplings[0] = a2dp_lhdc_samplings,
+	.samplings_size[0] = ARRAYSIZE(a2dp_lhdc_samplings),
+};
+
+struct a2dp_codec a2dp_lhdc_source = {
+	.dir = A2DP_SOURCE,
+	.codec_id = A2DP_CODEC_VENDOR_LHDC_V3,
+	.capabilities.lhdc_v3 = {
+		.info = A2DP_SET_VENDOR_ID_CODEC_ID(LHDC_V3_VENDOR_ID, LHDC_V3_CODEC_ID),
+		.frequency =
+			LHDC_SAMPLING_FREQ_44100 |
+			LHDC_SAMPLING_FREQ_48000 |
+			LHDC_SAMPLING_FREQ_96000,
+		.bit_depth =
+			LHDC_BIT_DEPTH_16 |
+			LHDC_BIT_DEPTH_24,
+		.jas = 0,
+		.ar = 0,
+		.version = LHDC_VER3,
+		.max_bit_rate = LHDC_MAX_BIT_RATE_900K,
+		.low_latency = 0,
+		.llac = 0, // TODO: copy LLAC/V3/V4 logic from AOSP patches
+		.ch_split_mode = LHDC_CH_SPLIT_MODE_NONE,
+		.meta = 0,
+		.min_bitrate = 0,
+		.larc = 0,
+		.lhdc_v4 = 1,
+	},
+	.capabilities_size = sizeof(a2dp_lhdc_v3_t),
+	.channels[0] = a2dp_lhdc_channels,
+	.channels_size[0] = ARRAYSIZE(a2dp_lhdc_channels),
+	.samplings[0] = a2dp_lhdc_samplings,
+	.samplings_size[0] = ARRAYSIZE(a2dp_lhdc_samplings),
+};
+
+void a2dp_lhdc_init(void) {
+}
+
+void a2dp_lhdc_transport_init(struct ba_transport *t) {
+
+	const struct a2dp_codec *codec = t->a2dp.codec;
+
+	if (codec->dir == A2DP_SINK) {
+		t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S24_4LE;
```

Comment:
Why sink and source use different bit-depth? From the A2DP config I can see that only 16 and 24 bits are supported...

Reply:
Decoder and encoder libraries have different APIs for producing/consuming samples, you may check headers for reference. I can change both to use S32_4LE, but internally encoder/decoder uses 16 or 24 bits depending on the configuration.